Gordon360/Services/RecIM/TeamService.cs

```diff
@@ -0,0 +1,511 @@
+                Subject = $"Gordon Rec-IM: {inviter.FirstName} {inviter.LastName} has invited you to a team!",
+                Body = messageBody,
+            };
+            message.IsBodyHtml = true;
+
+            smtpClient.Send(message);
+        }
+        
+        public async Task<ParticipantTeamViewModel> AddParticipantToTeamAsync(int teamID, ParticipantTeamUploadViewModel participant, string? inviterUsername = null)
+        {
+            //new check for enabling non-recim participants to be invited
+            if(!_context.Participant.Any(p => p.Username == participant.Username))
+            {
+                await _participantService.PostParticipantAsync(participant.Username, 1); //pending user
+            }
+
+
+            var participantTeam = new ParticipantTeam
+            {
+                TeamID = teamID,
+                ParticipantUsername = participant.Username,
+                SignDate = DateTime.Now,
+                RoleTypeID = participant.RoleTypeID ?? 2, //3 -> Member, 2-> Requested Join
+            };
+            await _context.ParticipantTeam.AddAsync(participantTeam);
+            await _context.SaveChangesAsync();
+            if (participant.RoleTypeID == 2 && inviterUsername is not null) //if this is an invite, send an email
+            {
+                await SendInviteEmail(teamID, participant.Username, inviterUsername);
+            }
+            return participantTeam;
+        }
+
+        public bool HasUserJoined(int activityID, string username)
+        {
+            // get all the partipantTeam from the teams with the activityID
+            var participantTeams = _context.Team.Where(t => t.ActivityID == activityID)
+                .Join(_context.ParticipantTeam.Where(pt => pt.RoleTypeID % 6 > 2),
+                t => t.ID,
+                pt => pt.TeamID,
+                (t, pt) => pt)
+                .AsEnumerable();
+
+            return participantTeams.Any(pt => pt.ParticipantUsername == username);
+        }
+
+        public bool HasTeamNameTaken(int activityID, string teamName)
+        {
+            return _context.Team.Any(t =>
+                        t.ActivityID == activityID
+                        && t.Name == teamName
+            );
+        }
+
+        public bool IsTeamCaptain(string username, int teamID)
+        {
+            return _context.ParticipantTeam.Any(t =>
+                        t.TeamID == teamID
+                        && t.ParticipantUsername == username
+                        && (
+                            t.RoleTypeID == 5       // RoleType: 5 => captain
+                            || t.RoleTypeID == 4    // RoleType: 4 => co-captain
+                        )
+            );
+        }
+
+        public int GetTeamActivityID(int teamID)
+        {
+            return _context.Team.FirstOrDefault(t => t.ID == teamID).ActivityID;
+        }
+
+        public int NumberOfGamesParticipatedByParticipant(int teamID, string username)
```

Comment:
```suggestion
        public int ParticipantAttendanceCount(int teamID, string username)
```
?

Reply:
What a funny function name! I think we should make it longer. I propose NumberOfMatchesOrGamesThatAParticipantParticipatedInByParticipantUsername